src/daft-logical-plan/src/optimization/rules/reorder_joins/brute_force_join_order.rs

```diff
@@ -0,0 +1,395 @@
+                    // divided by the selectivity of the join.
+                    // Assuming that join keys are uniformly distributed and independent, the selectivity is computed as the reciprocal
+                    // of the product of the largest total domains that form a minimum spanning tree of the relations.
+
+                    // TODO(desmond): This is a hack to get the selectivity of the join. We should expand this to take the minimum spanning tree
+                    // of edges to connect the relations. For simple graphs (e.g. the queries in the TPCH benchmark), taking the max of the total
+                    // domains is a good proxy.
+                    let denominator = connections
+                        .iter()
+                        .map(|conn| conn.total_domain)
+                        .max()
+                        .expect("There should be at least one total domain");
+                    let cardinality = left_join_order_tree.get_cardinality()
+                        * right_join_order_tree.get_cardinality()
+                        / denominator;
+                    // The cost of the join is the sum of the cardinalities of the left and right subgraphs, plus the cardinality of the joined graph.
+                    let cur_cost = cardinality + left_cost + right_cost;
+                    // Take the join with the lowest summed cardinality.
+                    if let Some(ref mut cur_min_cost) = min_cost {
+                        if *cur_min_cost > cur_cost {
+                            *cur_min_cost = cur_cost;
+                            *chosen_plan = Some(left_join_order_tree.join(
+                                right_join_order_tree,
+                                connections,
+                                cardinality,
+                            ));
+                        }
+                    } else {
+                        *min_cost = Some(cur_cost);
+                        *chosen_plan = Some(left_join_order_tree.join(
+                            right_join_order_tree,
+                            connections,
+                            cardinality,
+                        ));
+                    }
+                }
+            }
+            return;
+        }
+        if pick_idx >= elements.len() {
+            return;
+        }
+        // Case 1: Include the current element.
+        elements.swap(cur_idx, pick_idx);
+        Self::evaluate_combinations(
+            graph,
+            elements,
+            cur_idx + 1,
+            pick_idx + 1,
+            k_to_pick,
+            min_cost,
+            chosen_plan,
+        );
+        elements.swap(cur_idx, pick_idx); // Backtrack.
+
+        // Case 2: Exclude the current element in the chosen set.
+        Self::evaluate_combinations(
+            graph,
+            elements,
+            cur_idx,
+            pick_idx + 1,
+            k_to_pick,
+            min_cost,
+            chosen_plan,
+        );
```

Comment:
I'm not sure I understand this recursion here. Wouldn't case 2 evaluate the same split as the original call? Might need to just like visually walk through this with you to understand it

Reply:
Chatted offline about this